build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1188,6 +1212,34 @@ message UpdateActionResultRequest {
   ResultsCachePolicy results_cache_policy = 4;
 }
 
+// A request message for
+// [RepositoryCache.Download][build.bazel.remote.execution.v2.RepositoryCache.Download].
+message DownloadRequest {
+  // The instance of the execution system to operate against. A server may
+  // support multiple instances of the execution system (with their own workers,
+  // storage, caches, etc.). The server MAY require use of this field to select
+  // between them in an implementation-defined fashion, otherwise it can be
+  // omitted.
+  string instance_name = 1;
+
+  // List of mirror URLs referencing the same file.
+  repeated string urls = 2;
+
+  // Expected checksum of the file downloaded, in Subresource Integrity format.
+  string integrity = 3;
+
+  // If set, restrict cache hits to those cases where the file was added to the
+  // cache with the same canonical id
+  string canonical_id = 4;
+}
+
+// A response message for
+// [RepositoryCache.Download][build.bazel.remote.execution.v2.RepositoryCache.Download].
+message DownloadResponse {
+  // The blob digest of the downloaded file.
+  Digest digest = 1;
```

Comment:
In a similar vein, this assumes the downloaded asset can be expressed as a single file, and that that's the desired format. Is that a reasonable assumption?

I personally believe we'll want to support packed and unpacked outputs before too long - e.g. for URLs that point at tarballs, the Digest of the tarball itself, and/or some index of its contents pre-unpacked.

Would it then also make sense to symmetrically express this as `repeated google.protobuf.Any contents = 1`, and enable the server to support any format(s) it wishes?

I'd suggest also privileging Digest (e.g. making it its own field, with a second field for alternates), except I don't know what to do with the case where the contents are not reasonably expressed by a single digest. For that reason, I suggest we mention Digest in comments but *not* require that every download include a single Digest amongst its response types. Does that make sense?

Reply:
This API is specifically designed for the single-file use case, such as Bazel's `repository_ctx.download()` or Buck's `http_archive()`. I want to use it as a distributed version of Bazel's `--repository_cache` flag.

There are potentially many ways that new caching behavior could be added in the future, but I think that discussion belongs in separate proposals and separate RPCs.